### PR TITLE
[#1336] Demo extension

### PR DESCRIPTION
**Pull request by SilviaAmAm**
Part of #1336

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -157,6 +157,23 @@ jobs:
           path: image.tar
           retention-days: 1
 
+      - name: Build the test Docker image
+        run: |
+          docker build . \
+            --tag $IMAGE_NAME:test \
+            --build-arg ENVIRONMENT=ci \
```
Comment (sergei-maertens) [inline .github/workflows/ci.yml:164]: the Dockerfile specifies `ARG ENVIRONMENT_SETTINGS`, while you use `ENVIRONMENT` here. I'd opt for the shorter one.

File: Dockerfile [hunk 2]
```diff
@@ -4,12 +4,14 @@
 # image minimal in size.
 
 # must be at the top to use it in FROM clauses
+ARG ENVIRONMENT_SETTINGS=production
```
Comment (sergei-maertens) [inline Dockerfile:7]: this is only needed for the backend-build stage, so I'd move it over there (with the default), close to the `pip install` line (around 42)

File: docs/developers/extending.rst [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+* Authentication
+* Prefill
+* Payments
+* Appointments
+* Formio formatters
+* Formio custom types
+* Validations
+
+Modules and plugins are Python packages with some Django bindings. They are called
+by the Open Forms core functionality (forms, submissions...).
```
Comment (sergei-maertens) [inline docs/developers/extending.rst:22]: This is a lot of repetition from `developers/architecture.rst`, so I'd reference that instead. Also, the technial documentation about that needs to be expanded upon, but that's another ticket.

File: docs/developers/extending.rst [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+   Instructions on how to implement such 'external plugins' are described below.
+
+.. _CONTRIBUTING.md: https://github.com/open-formulieren/open-forms/blob/master/CONTRIBUTING.md
+
+Creating an external plugin
+===========================
+
+The `demo extension <https://github.com/open-formulieren/demo-extension>`_ is an example of an extension plugin.
+
+An extension must be implement as a Django package, which is a stricter form of a
```
Comment (sergei-maertens) [inline docs/developers/extending.rst:39]:
```suggestion
An extension must be implemented as a Django package, which is a stricter form of a
```

File: docs/developers/extending.rst [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+   If you add a solo model, you will have to generate migrations.
+   See the :ref:`section below<test_extension_plugin_locally>` for details.
+
+#. For certain types of plugins, like for registration backends, it is also possible to add form-specific options.
+   These can be specified with a serializer. See `config.py`_ for details.
+
+If any initialization or module loading is needed, this should be done as part of the
+extension's ``extension.apps.ExtensionConfig.ready`` hook, which is called when Django
+bootstraps. This is the mechanism by which Open Forms plugins are registered
+(refer to the :ref:`plugin section<plugins_index>` for further details).
```
Comment (sergei-maertens) [inline docs/developers/extending.rst:78]: this looks copied a bit too-verbatim from the issue ;)

File: docs/developers/extending.rst [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+Django.
+
+If you need to generate migrations for your package, you can now do it as follows (from within the Open Forms directory):
+
+.. code-block:: bash
+
+    python src/manage.py makemigrations demo_extension
+    python src/manage.py migrate
+
+If you created a solo model, you can add the configuration page to the Admin. To do this, login into the Open Forms
```
Comment (sergei-maertens) [inline docs/developers/extending.rst:114]:
```suggestion
If you created a solo model, you can add the configuration page to the Admin. To do this, log in into the Open Forms
```
?

File: docs/developers/extending.rst [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+Admin:
+
+#. Go to the **Configuratie** > **Application groups**.
+
+#. Click on **Configuratie**.
+
+#. In the **Models** section, look for the name of your configuration model in the left table (for the demo extension, this was  ``demo_extension.Demoextensionconfig``).
+   Then double click on it to add it to the right table.
+
+#. Save the configuration.
```
Comment (sergei-maertens) [inline docs/developers/extending.rst:124]: We'll have to find a better solution for this, as I think the post-migrate signal handlers deletes any manual changes and restores the fixture configuration.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]
The extending documentation is a bit too specific in describing this particular plugin. I'd rather expect a more generic/broader documentation describing the principles, while referencing the demo plugin for a concrete example.

File: docs/developers/extending.rst (hunk 5 above)
Comment (SilviaAmAm) [reply docs/developers/extending.rst:78]: :sweat_smile:

File: docs/developers/extending.rst [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+Creating an external plugin
+===========================
+
+An extension must be implemented as a Django package, which is a stricter form of a
+regular Python package. They can be developed independently from the open-formulieren Github organisation.
+
+The extensions can be loaded at deployment time through the environment variable ``OPEN_FORMS_EXTENSIONS`` which
+specifies the Python name of the extensions to load. The Open Forms ``base.py`` settings module will extract this
+configuration value and append the list of packages to ``INSTALLED_APPS``, which causes them to be loaded by
+Django. For an extension to be loaded, it needs to be present in the ``PYTHONPATH``.
```
Comment (sergei-maertens) [inline docs/developers/extending.rst:30]:
```suggestion
specifies the Python name of the extensions to load. Open Forms will extract this configuration value and 
load the referenced extension packages. For an extension to be loaded, it needs to be present in the ``PYTHONPATH``.
```
Simplified a bit without mentioning too many implementation details.

File: docs/developers/extending.rst [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+This means that it needs to be either:
+
+* In the ``src/`` directory of Open Forms. It is then automatically picked up.
+* Anywhere on the file path, but the ``PYTHONPATH`` is modified to include the path to the extension.
```
Comment (sergei-maertens) [inline docs/developers/extending.rst:34]:
```suggestion
* Anywhere on the file path, but the ``PYTHONPATH`` environment variable is modified to include the path to the extension.
```

File: docs/developers/extending.rst [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+* In the relevant site-packages directory, similarly to when an Open Forms dependency is installed with ``pip install``
+   and ``virtualenv``.
+
+To create a Docker image that extends an Open Forms release, a custom ``Dockerfile`` can be created and the extension
+source code can be added to the desired location. Another option is to mount the source code as a volume in a custom
+``docker-compose.yml`` file.
```
Comment (sergei-maertens) [inline docs/developers/extending.rst:40]:
```suggestion
Building and distributing the extended Open Forms
-------------------------------------------------
You can build a custom Docker image extending an Open Forms release with a custom `Dockerfile`, in which the
extension source code is added to the desired location. Another option is to mount the source code as a volume 
in a custom ``docker-compose.yml`` file or as part of your Kubernetes manifests, depending on how you deploy your instance(s).
```

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

**Review by sergei-maertens — APPROVED** [review]

**Pull request merged**